torchtune/training/_distributed.py

```diff
@@ -242,7 +245,9 @@ def gather_cpu_state_dict(
         if sharded_param.is_cpu:
             # Move back to device if offloaded to CPU
             sharded_param = sharded_param.to(device)
-        if isinstance(sharded_param._local_tensor, NF4Tensor):
+        if hasattr(sharded_param, "_local_tensor") and isinstance(
```

Comment:
If the tensor isn't sharded but it still a NF4Tensor, we still need to upcast the datatype as is done on line 271

Reply:
so maybe another `elif` in the case that the tensor is unsharded, but still NF4Tensor, and just need to call line 271?

Reply:
I think this would be much cleaner if at the top of the loop we do:
```
if hasattr(...):
	# get full tensor (NF4 or .full_tensor)
if full_param is NF4:
	# upcast
if is_rank_zero:
	# the rest the same
```
This code still seems to assume that the unsharded param is on rank0 which isn't guaranteed